server/src/main/java/org/elasticsearch/indices/recovery/RemoteRecoveryTargetHandler.java

```diff
@@ -174,15 +210,64 @@ public void writeFileChunk(StoreFileMetadata fileMetadata, long position, BytesR
             throttleTimeInNanos = 0;
         }
 
-        transportService.sendRequest(targetNode, PeerRecoveryTargetService.Actions.FILE_CHUNK,
-            new RecoveryFileChunkRequest(recoveryId, shardId, fileMetadata, position, content, lastChunk,
-                totalTranslogOps,
-                /* we send estimateTotalOperations with every request since we collect stats on the target and that way we can
-                 * see how many translog ops we accumulate while copying files across the network. A future optimization
-                 * would be in to restart file copy again (new deltas) if we have too many translog ops are piling up.
-                 */
-                throttleTimeInNanos), fileChunkRequestOptions, new ActionListenerResponseHandler<>(
-                    ActionListener.map(listener, r -> null), in -> TransportResponse.Empty.INSTANCE, ThreadPool.Names.GENERIC));
+        final TimeValue timeout = fileChunkRequestOptions.timeout();
+        /* we send estimateTotalOperations with every request since we collect stats on the target and that way we can
+         * see how many translog ops we accumulate while copying files across the network. A future optimization
+         * would be in to restart file copy again (new deltas) if we have too many translog ops are piling up.
+         */
+        final RecoveryFileChunkRequest request = new RecoveryFileChunkRequest(recoveryId, shardId, fileMetadata,
+            position, content, lastChunk, totalTranslogOps, throttleTimeInNanos);
+
+        Consumer<ActionListener<Void>> action = actionListener ->
+            transportService.sendRequest(targetNode, PeerRecoveryTargetService.Actions.FILE_CHUNK, request, fileChunkRequestOptions,
+                new ActionListenerResponseHandler<>(ActionListener.map(actionListener, r -> null),
+                    in -> TransportResponse.Empty.INSTANCE, ThreadPool.Names.GENERIC));
+
+        executeRetryableAction(action, listener, timeout);
     }
 
+    @Override
+    public void cancel() {
```

Comment:
I wonder if we should make all these changes in `RecoverySourceHandler` instead? It might be simpler than doing it here.

Reply:
The advantage there is that the cancellation logic is in one place. The advantage here is that it's clear that the retries are due to inter-node communication, and might eventually be folded into some TransportService infrastructure?

Reply:
I left as is for now.